M2/Macaulay2/packages/ResLengthThree.m2

```diff
@@ -0,0 +1,752 @@
+newPackage ( "ResLengthThree",
+    Version => "1.0",
+    Date => "16 November 2020",
+    Authors => {
+	{ Name => "Lars Winther Christensen",
+	  Email => "lars.w.christensen@ttu.edu",
+	  HomePage => "http://www.math.ttu.edu/~lchriste/index.html" },
+      { Name => "Luigi Ferraro",
+	  Email => "lferraro@ttu.edu",
+	  HomePage => "http://www.math.ttu.edu/~lferraro" },
+	{ Name => "Francesca Gandini",
+	  Email => "fra.gandi.phd@gmail.com",
+	  HomePage => "https://github.com/fragandi" },
+	{ Name => "Frank Moore",
+	  Email => "moorewf@wfu.edu",
+	  HomePage => "http://users.wfu.edu/moorewf/" },
+	{ Name => "Oana Veliche", 
+	  Email => "o.veliche@northeastern.edu",
+	  HomePage => "https://web.northeastern.edu/oveliche/index.html" }
+	},
+    Headline => "Multiplication in free resolutions of length three",
+    Reload => false,
+    DebuggingMode => false
+    )
+
+export { "resLengthThreeAlg", "resLengthThreeTorAlg", "multTableOneOne", "multTableOneTwo", 
+    "resLengthThreeTorAlgClass", "makeRes", "Labels", "Compact" }
+
+
+--==========================================================================
+-- EXPORTED FUNCTIONS
+--==========================================================================
+
+resLengthThreeAlg = method()
+
+resLengthThreeAlg(ChainComplex, List) := (F, sym) -> (
+   if F.cache#?"Algebra Structure" then return F.cache#"Algebra Structure";
+   if length F != 3 then
+     error "Expected a chain complex of length three which is free of rank one in degree zero.";
+   if #sym != 3 or any(sym, s -> (class baseName s =!= Symbol))  then
+     error "Expected a list of three symbols.";
+   mult := multTables(F);
+   Q := ring F;
+   m := numcols F.dd_1;
+   l := numcols F.dd_2;
+   n := numcols F.dd_3;        
+   degreesP := if isHomogeneous F then 
+                  flatten apply(3, j -> apply(degrees source F.dd_(j+1), d -> {j+1} | d))
+	       else
+	          flatten apply(3, j -> apply(degrees source F.dd_(j+1), d -> {0} | d));
+   skewList := toList((0..(m-1)) | ((m+l)..(m+l+n-1)));
+   e := baseName (sym#0);
+   f := baseName (sym#1);
+   g := baseName (sym#2);
+   -- use this line if you want to ensure that 'basis' works properly on the returned ring.
+   --P := first flattenRing (Q[e_1..e_m,f_1..f_l,g_1..g_n,SkewCommutative=>skewList, Degrees => degreesP, Join => false]);
+   P := Q[e_1..e_m,f_1..f_l,g_1..g_n,SkewCommutative=>skewList, Degrees => degreesP, Join => false];
```

Comment:
This will set values for the user's variables, e, f, and g, which might surprise the user.  Better would be to change `Q[...]` to `Q(monoid[...])` to avoid that.  The user can always use `use` to get those variables set.